File: packages/rust/fluster_bibliography/src/data/embedded_csl_file.rs
use serde::{Deserialize, Serialize};
use strum::EnumIter;

#[derive(uniffi::Enum, strum_macros::Display, EnumIter, Clone, Serialize, Deserialize)]
pub enum EmbeddedCslFile {
    #[serde(rename = "american-chemical-society.csl")]
    #[strum(to_string = "american-chemical-society.csl")]
    AmericanChemicalSociety,
    #[serde(rename = "american-institute-of-physics.csl")]
    #[strum(to_string = "american-institute-of-physics.csl")]
    AmericanInstituteOfPhysics,
    #[serde(rename = "american-medical-association.csl")]
    #[strum(to_string = "american-medical-association.csl")]
    AmericanMedicalAssociation,
    #[serde(rename = "american-political-science-association.csl")]
    #[strum(to_string = "american-political-science-association.csl")]
    AmericanPoliticalScienceAssociation,
    #[serde(rename = "american-sociological-association.csl")]
    #[strum(to_string = "american-sociological-association.csl")]
    AmericanSociologicalAssociation,
    #[serde(rename = "apa.csl")]
    #[strum(to_string = "apa.csl")]
    Apa,
    #[serde(rename = "association-for-computing-machinery.csl")]
    #[strum(to_string = "association-for-computing-machinery.csl")]
    AssociationForComputingMachinery,
    #[serde(rename = "cell.csl")]
    #[strum(to_string = "cell.csl")]
    Cell,
    #[serde(rename = "chicago-author-date.csl")]
    #[strum(to_string = "chicago-author-date.csl")]
    ChicagoAuthorDate,
    #[serde(rename = "chicago-notes-bibliography.csl")]
    #[strum(to_string = "chicago-notes-bibliography.csl")]
    ChicagoNotesBibliography,
    #[serde(rename = "harvard-cite-them-right.csl")]
    #[strum(to_string = "harvard-cite-them-right.csl")]
    HarvardCiteThemRight,
    #[serde(rename = "ieee.csl")]
    #[strum(to_string = "ieee.csl")]
    Ieee,
    #[serde(rename = "modern-language-association.csl")]
    #[strum(to_string = "modern-language-association.csl")]
    ModernLanguageAssociation,
    #[serde(rename = "nature.csl")]
    #[strum(to_string = "nature.csl")]
    Nature,
    #[serde(rename = "science.csl")]
    #[strum(to_string = "science.csl")]
    Science,
    #[serde(rename = "springer-vancouver.csl")]
    #[strum(to_string = "springer-vancouver.csl")]
    SpringerVancouver,
}

File: packages/swift/FlusterData/Sources/FlusterData/constants/embedded_csl_file.swift
//
//  File.swift
//  FlusterData
//
//  Created by Andrew on 2/21/26.
//

import Foundation

/// A wrapper around the EmbeddedCslFile enum provided from rust. The same exact values with some utility methods and some parent classes.
public enum EmbeddedCslFileSwift: String, CaseIterable, Hashable {
  case americanChemicalSociety = "American Chemical Society"
  case americanInstituteOfPhysics = "American Institute of Physics"
  case americanMedicalAssociation = "American Medical Association"
  case americanPoliticalScienceAssociation = "American Political Science Association"
  case americanSociologicalAssociation = "American Sociological Association"
  case apa = "APA"
  case associationForComputingMachinery = "Association for Computing Machinery"
  case cell = "Cell"
  case chicagoAuthorDate = "Chicago: Author, Date"
  case chicagoNotesBibliography = "Chicago: Notes bibliography"
  case harvardCiteThemRight = "Harvard: Cite them right"
  case ieee = "Ieee"
  case modernLanguageAssociation = "Modern Language Association"
  case nature = "Nature"
  case science = "Science"
  case springerVancouver = "Springer: Vancouver"

  func toEmbeddedFilePath() -> String {
    switch self {
      case .americanChemicalSociety:
        return "american-chemical-society"
      case .americanInstituteOfPhysics:
        return "american-institute-of-physics"
      case .americanMedicalAssociation:
        return "american-medical-association"
      case .americanPoliticalScienceAssociation:
        return "american-political-science-association"
      case .americanSociologicalAssociation:
        return "american-sociological-association"
      case .apa:
        return "apa"
      case .associationForComputingMachinery:
        return "association-for-computing-machinery"
      case .cell:
        return "cell"
      case .chicagoAuthorDate:
        return "chicago-author-date"
      case .chicagoNotesBibliography:
        return "chicago-notes-bibliography"
      case .harvardCiteThemRight:
        return "harvard-cite-them-right"
      case .ieee:
        return "ieee"
      case .modernLanguageAssociation:
        return "modern-language-association"
      case .nature:
        return "nature"
      case .science:
        return "science"
      case .springerVancouver:
        return "springer-vancouver"
    }
  }

  public func toFlusterBibliographyCslFile() -> String {
        let fp = self.toEmbeddedFilePath()
        if let fileURL = Bundle.main.url(forResource: "csl/\(fp)", withExtension: "csl", subdirectory: "bibliography_embedded") {
            do {
                let s = try String(contentsOf: fileURL)
                return s
            } catch {
                print("Error: \(error.localizedDescription)")
                fatalError("Failed to get embedded csl file.")
            }
        } else {
            // The file was not found
            print("File not found.")
            fatalError("Failed to get embedded csl file.")
        }
      // This should never be reached. Make sure to throw fatal error if the embedded file can't be found.
      return ""
  }
}


public func getCslLocaleFileContent() -> String {
    if let fileURL = Bundle.main.url(forResource: "csl_locale/en_us", withExtension: "xml", subdirectory: "bibliography_embedded") {
        do {
            let s = try String(contentsOf: fileURL)
            return s
        } catch {
            print("Error: \(error.localizedDescription)")
            fatalError("Failed to get embedded csl locale file.")
        }
    } else {
        fatalError("Failed to get embedded csl locale file.")
    }
  // This should never be reached. Make sure to throw fatal error if the embedded file can't be found.
  return ""
}
